kernel/src/platform/mpu.rs

```diff
@@ -257,3 +267,70 @@ pub trait MPU {
 
 /// Implement default MPU trait for unit.
 impl MPU for () {}
+
+/// The generic trait that particular kernel memory protection unit
+/// implementations need to implement.
+///
+/// This trait provides generic functionality to extend the MPU trait above
+/// to also allow the kernel to protect itself. It is expected that only a
+/// limited number of SoCs can support this, which is why it is a seperate
+/// implementation.
+pub trait KernelMPU {
+    /// MPU-specific state that defines a particular configuration for the kernel
+    /// MPU.
+    /// That is, this should contain all of the required state such that the
+    /// implementation can be passed an object of this type and it should be
+    /// able to correctly and entirely configure the MPU.
+    ///
+    /// It is `Default` so we can create empty state when the process is
+    /// created, and `Display` so that the `panic!()` output can display the
+    /// current state to help with debugging.
+    type KernelMpuConfig: Default + Display = MpuConfigDefault;
+
+    /// Mark a region of memory that the Tock kernel owns
+    ///
+    /// This function will optionally set the MPU to enforce the specified
+    /// constraints for all accessess (even from the kernel).
+    /// This should be used to mark read/write/execute only areas of the Tock
+    /// kernel to have the hardware enforce those permissions.
+    ///
+    /// If using this function, all kernel executable code must be marked as
+    /// executable prior to enabling the MPU (by calling `enable_mpu()).
+    /// Also all kernel read/write regions must be marked. That is any
+    /// unmarked kernel regions will cause a fault if accesses.
+    ///
+    /// Not all architectures support this, so don't assume this will be
+    /// implemented.
+    ///
+    /// # Arguments
+    ///
+    /// - `memory_start`:             start of memory region
+    /// - `memory_size`:              size of unallocated memory
+    /// - `permissions`:              permissions for the region
+    /// - `config`:                   MPU region configuration
+    ///
+    /// # Return Value
+    ///
+    /// Returns the start and size of the requested memory region. If it is
+    /// infeasible to allocate the MPU region, returns None.
+    #[allow(unused_variables)]
+    fn allocate_kernel_region(
+        &self,
+        memory_start: *const u8,
+        memory_size: usize,
+        permissions: Permissions,
+        config: &mut Self::KernelMpuConfig,
+    ) -> Option<Region> {
+        None
+    }
+
+    /// Enables the MPU for the kernel.
+    ///
+    /// This function must enable the permission restrictions on the various
```

Comment:
I don't understand the semantics of this operation. Aren't these regions disabled while in userspace? The comment "cannot be disabled" confuses me.

Reply:
No, these will always be enabled.

The idea is that the kernel can specify that this region of memory is read/write/execute for the kernel. All other accesses are blocked, which includes user space. Once enabled this can't be disabled by either the kernel or the app.

The kernel should allow itself access to what it needs, then lock out everything else from itself and apps.